add tests

giusbianco opened the pull request:
At the moment, the project is in a functional state and meets the requirements. However, I still need to streamline the library by simplifying the code and removing unnecessary elements. This will improve the tool's efficiency and ensure it is more user-friendly. I'm working on this final step to deliver a polished and optimized final product.

giusbianco committed:
added unit tests

Work is not complete, but needed this push to understand current level of code coverage. Unit tests added to binary (ci-generate.rs) and library (lib.rs).

giusbianco committed:
added unit tests - library and binary

giusbianco committed:
Add tests

corrected code formatting

giusbianco committed:
edit/add tests

giusbianco committed:
Update ci-generate.rs

giusbianco committed:
Update lib.rs

giusbianco committed:
add unit tests

giusbianco committed:
add unit tests

giusbianco committed:
Update lib.rs

giusbianco committed:
Update lib.rs

giusbianco committed:
Update tests on lib.rs

giusbianco committed:
Update tests on lib.rs

giusbianco committed:
Update test

giusbianco committed:
add/edit tests

added tests in filter.rs and ignored some tests in lib.rs

giusbianco committed:
add test

giusbianco committed:
add integration tests

Several changes in this commit:
1. adding path validation functionality for ubuntu;
2. integration of the functionality into the various templates, with appropriate code changes;
3. integration testing for each template;
3.1. creation of output folder to hold projects;
3.2. hash calculation for each project file and comparison with expected hash.

giusbianco committed:
add tests

Several changes in this commit:
1. adding path validation functionality for ubuntu;
2. integration of the functionality into the various templates, with appropriate code changes;
3. integration testing for each template;
3.1. creation of output folder to hold projects;
3.2. hash calculation for each project file and comparison with expected hash.
4. removed unit tests in the binary

giusbianco committed:
edit tests

Several changes in this commit:
1. adding path validation functionality for ubuntu;
2. integration of the functionality into the various templates, with appropriate code changes;
3. integration testing for each template;
3.1. creation of output folder to hold projects;
3.2. hash calculation for each project file and comparison with expected hash.
4. removed unit tests in the binary

giusbianco committed:
Update integration_test.rs

giusbianco committed:
Update Cargo.toml

giusbianco committed:
add windows path validation

giusbianco committed:
Update mod.rs

giusbianco committed:
Update tests

Several changes in this commit:
1. adding path validation functionality for ubuntu;
2. integration of the functionality into the various templates, with appropriate code changes;
3. integration testing for each template;
3.1. creation of output folder to hold projects;
3.2. hash calculation for each project file and comparison with expected hash.
4. removed unit tests in the binary

giusbianco committed:
Update

giusbianco committed:
Update integration tests

giusbianco committed:
add integration tests

added a full suite of integration tests, provided with example templates and use of the configuration file. Also implemented the insta snapshot function for making assertions.

giusbianco committed:
Add insta tests

added snapshot tests for both templates generated with default values + cli, and templates generated with custom configuration file

giusbianco committed:
Update README.md

giusbianco committed:
Update

giusbianco committed:
Update library test

giusbianco committed:
Update README.md

Luni-4 commented on README.md:
```diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
 
+## Testing
+
+In ci-generate we use [insta] (https://insta.rs), a library of snapshots tests for Rust.
```
Without `ci-generate`, we can start with `We use...`

Luni-4 commented on README.md:
```diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+## Testing
+
+In ci-generate we use [insta] (https://insta.rs), a library of snapshots tests for Rust.
+The first time the command 
```
`To run tests, launch the following command:`

Luni-4 commented on README.md:
````diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+```
+
+When it is executed, snaps are created with the current content of the files being tested, which will be considered as basic snapshots. 
+Once the basic snapshots have been created, you can run the previous command again to perform a content test by producing new snapshots that will then be compared with the basic snapshots.
````
I will explain how to accept snapshots, the procedure with either the `--accept` option or `cargo insta accept` command, and then introduce the concept of basic snapshots. So first the procedure, saying that tests should fail the first time, and then the definition of concepts. In this way we also explain how we arrived to the concept of snapshots

Luni-4 commented on README.md:
```diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+
+### Updating insta tests
+
+When the comparison is made between snaps, tests return a **failure** if there are changes from the base snap, or return a **success** if the two snaps coincide. 
```
`When a comparison between snaps is made, a test returns...` `otherwise a success...`

Luni-4 commented on README.md:
````diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+cargo insta review --include-hidden
+```
+
+will allow you to review the differences between the two snaps and possibly accept the changes. The `--include-hidden` option is important to add in this context because it allows you to verify the snap of hidden files or folders, present in template creation for example.
````
`It is important to add the `--include-hidden` option in this context`

Luni-4 commented on README.md:
```diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+
+will allow you to review the differences between the two snaps and possibly accept the changes. The `--include-hidden` option is important to add in this context because it allows you to verify the snap of hidden files or folders, present in template creation for example.
+
+With this in mind, it is possible to test a series of snapshots of different contents.
```
`All of this allows to test....`

Luni-4 commented on README.md:
```diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+
+#### Configuration files
+
+It is possible to test the content of a custom configuration `.toml` file. Refer to the section [Configuration](#configuration) to set up a tamplate with personal `config.toml`. For example, it is possible to use:
```
`There is also the possibility to test...`

`template with your own `config.toml`

Luni-4 commented on README.md:
````diff
@@ -95,6 +95,49 @@ $ ci-generate meson -l LGPL-2.1
+``` sh
+ci-generate -c tests/repositories/config_template/config.toml meson -l=APL-1.0 -b=master tests/repositories/config_template/meson_template_config
+```
+to create a meson template with a previously created configuration file.
````
`...configuration file created previously`

Luni-4 commented on src/lib.rs:
```diff
@@ -61,7 +61,10 @@ impl CiTemplate {
+            match create_dir_all(dir) {
+                Ok(x) => x,
+                _ => return Err(Error::NoDirExists),
+            }
```
Can we reduce the code using `create_dir_all(dir)?`? It we can, let's replace this pattern everywhere

Luni-4 commented on src/lib.rs:
```diff
@@ -70,19 +73,20 @@ impl CiTemplate {
+                Ok(x) => x,
+                _ => return Err(Error::NoContext),
+            };
+            let _ = write(path, filled_template);
```
Is it necessary the `_`? Is it a `clippy` warning lint?

Luni-4 commented on tests/cargo_test.rs:
```diff
@@ -0,0 +1,30 @@
+const SNAPSHOT_PATH_C: &str = "../repositories/snapshots/config/cargo/";
+
+const CONFIG_R: &str = "tests/repositories/config_template/config.toml";
+const CONFIG_S: &str = "../repositories/snapshots/config/";
```
We can convert them in `Path` instead of `&str`

Luni-4 commented on tests/cargo_test.rs:
```diff
@@ -0,0 +1,30 @@
+// $ ci-generate cargo --docker-image-description=docker_image --license=EUPL-1.2 --name=Project --branch=main tests/repositories/cargo_template
+#[test]
+fn test_cargo() {
+    compare_template_output_with_expected_one(&PathBuf::from(SNAPSHOT_PATH), &PathBuf::from(REPO));
```
We can define `<P: AsRef<Path>` as generic, in this way we can pass both `PathBuf` and `Path`

Luni-4 reviewed (changes requested):
Thanks a lot for your hard work! Just a series of changes

giusbianco committed:
Update integration tests

giusbianco committed:
Update path_validation

giusbianco committed:
Update tests

giusbianco committed:
Update tests

Currently, we are working on implementing an API for simplifying library functionalities, which includes the CommonData, MavenData, MesonData, and CargoData structures. This implementation is still in progress and is aimed at facilitating testing without the need for a separate {config.toml} file to define parameters to be passed to templates. By doing so, we aim to streamline the testing process and make it more efficient.

giusbianco committed:
Update format file

giusbianco committed:
Update README.md

giusbianco committed:
Windows adjustments

giusbianco committed:
Update README.md

giusbianco committed:
Update README.md

giusbianco committed:
Edit path_validation()

Removed the error type: PathNotExist

giusbianco committed:
Edit details

giusbianco committed:
Add temporary directory

giusbianco committed:
Merge branch 'add-test' of https://github.com/SoftengPoliTo/ci-generate into add-test

giusbianco committed:
Update mod.rs

Simplify compare_template function

giusbianco committed:
Update

Update mod.rs:
Simplify compare_template function

giusbianco committed:
Update

Update mod.rs:
Simplify compare_template function

giusbianco committed:
Update mod.rs

Simplify compare_template function

giusbianco committed:
Update lib matches

giusbianco committed:
Update lib matches

giusbianco committed:
Update tests

Edit Templates:
Modified Maven and Cargo templates, it is now possible to call the appropriate function to set the base parameter.

Update tests:
Implementation of setter and getter, to keep fields of struct CommonData private. Fixed field call dependencies.

giusbianco committed:
Update tests

Edit Templates:
Modified Maven and Cargo templates, it is now possible to call the appropriate function to set the base parameter.

Update tests:
Implementation of setter and getter, to keep fields of struct CommonData private. Fixed field call dependencies.

giusbianco committed:
Edit Templates

Modified Maven and Cargo templates, it is now possible to call the appropriate function to set the base parameter.

giusbianco committed:
Edit Templates

Modified Maven and Cargo templates, it is now possible to call the appropriate function to set the base parameter.

giusbianco committed:
Update template_test

giusbianco committed:
Update template_test

giusbianco committed:
Update tests

Implementation of setter and getter, to keep fields of struct CommonData private. Fixed field call dependencies.

giusbianco committed:
Update tests

Implementation of setter and getter, to keep fields of struct CommonData private. Fixed field call dependencies.

giusbianco committed:
Update tests

giusbianco committed:
Update toolchain

improved code

giusbianco committed:
Update create_ci and create_project

giusbianco committed:
Update create_ci and create_project

giusbianco committed:
Update create_ci and create_project

giusbianco committed:
Update create_ci and create_project

giusbianco committed:
Update fmt

giusbianco committed:
Update fmt

giusbianco committed:
Edit variable name

giusbianco committed:
Edit Template fields

giusbianco committed:
Edit Template fields

giusbianco committed:
Update Cargo fields

giusbianco committed:
Update Cargo fields

giusbianco committed:
Merge branch 'add-test' of https://github.com/SoftengPoliTo/ci-generate into add-test

giusbianco committed:
Edit README

giusbianco committed:
Edit README

giusbianco committed:
Merge branch 'add-test' of https://github.com/SoftengPoliTo/ci-generate into add-test
